src/System.Net.WebSockets.Client/src/System/Net/WebSockets/WebSocketHandle.Unix.cs

```diff
@@ -770,144 +760,162 @@ private static void WriteRandomMask(byte[] buffer, int offset)
             {
-                while (true) // in case we get control frames that should be ignored from the user's perspective
+                CancellationTokenRegistration registration = cancellationToken.Register(s => ((ManagedClientWebSocket)s).Abort(), this);
+                try
                 {
-                    // Get the last received header.  If its payload length is non-zero, that means we previously
-                    // received the header but were only able to read a part of the fragment, so we should skip
-                    // reading another header and just proceed to use that same header and read more data associated
-                    // with it.  If instead its payload length is zero, then we've completed the processing of
-                    // thta message, and we should read the next header.
-                    MessageHeader header = _lastReceiveHeader;
-                    if (header.PayloadLength == 0)
-                    {
-                        MessageHeader? headerOpt = await ReadMessageHeaderAsync(cancellationToken).ConfigureAwait(false);
-                        if (headerOpt == null)
-                        {
-                            // The connection closed; nothing more to read.
-                            return new WebSocketReceiveResult(0, WebSocketMessageType.Text, true);
-                        }
-                        header = headerOpt.GetValueOrDefault();
-                    }
-
-                    // If the header represents a ping or a pong, handle it.
-                    if (header.Opcode == MessageOpcode.Ping || header.Opcode == MessageOpcode.Pong)
+                    while (true) // in case we get control frames that should be ignored from the user's perspective
                     {
-                        // Consume any (optional) payload associated with the ping/pong
-                        if (header.PayloadLength > 0 && _receiveBufferCount < header.PayloadLength)
+                        // Get the last received header.  If its payload length is non-zero, that means we previously
+                        // received the header but were only able to read a part of the fragment, so we should skip
+                        // reading another header and just proceed to use that same header and read more data associated
+                        // with it.  If instead its payload length is zero, then we've completed the processing of
+                        // thta message, and we should read the next header.
+                        MessageHeader header = _lastReceiveHeader;
+                        if (header.PayloadLength == 0)
                         {
-                            await EnsureBufferContainsAsync((int)header.PayloadLength, cancellationToken).ConfigureAwait(false);
-                        }
+                            if (_receiveBufferCount < MaxMessageHeaderLength && // fast check to see if we have enough data for any possible header
```

Comment:
Or... maybe these headers are different headers than I was expecting; since they seem to be length-prefixed values :).  But, still, the boundary case question applies; just to make sure we don't have an off-by-one error here.